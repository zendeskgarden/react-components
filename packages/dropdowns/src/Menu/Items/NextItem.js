/**
 * Copyright Zendesk, Inc.
 *
 * Use of this source code is governed under the Apache License, Version 2.0
 * found at http://www.apache.org/licenses/LICENSE-2.0.
 */

import React from 'react';
import PropTypes from 'prop-types';

import Item from './Item';
import { StyledNextItem } from '../../styled';
import useDropdownContext from '../../utils/useDropdownContext';
import useMenuContext from '../../utils/useMenuContext';

/**
 * Accepts all `<li>` props
 */
const NextItem = React.forwardRef(({ value, disabled, ...props }, ref) => {
  const { nextItemsHashRef } = useDropdownContext();
  const { itemIndexRef } = useMenuContext();

  if (disabled) {
    return <Item component={StyledNextItem} disabled {...props} />;
  }

  // Include current index in global Dropdown context
  nextItemsHashRef.current[value] = itemIndexRef.current;

  return (
    <Item component={StyledNextItem} aria-expanded={true} value={value} ref={ref} {...props} />
  );
});

NextItem.propTypes = {
  value: PropTypes.any,
  disabled: PropTypes.bool,
  active: PropTypes.bool,
  focused: PropTypes.bool,
  hovered: PropTypes.bool
};

export default NextItem;
