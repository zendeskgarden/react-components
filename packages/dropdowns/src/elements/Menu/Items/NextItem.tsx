/**
 * Copyright Zendesk, Inc.
 *
 * Use of this source code is governed under the Apache License, Version 2.0
 * found at http://www.apache.org/licenses/LICENSE-2.0.
 */

import React from 'react';
import PropTypes from 'prop-types';

import { Item, IItemProps } from './Item';
import { StyledNextItem, StyledItemIcon, StyledNextIcon } from '../../../styled';
import useDropdownContext from '../../../utils/useDropdownContext';
import useMenuContext from '../../../utils/useMenuContext';

const NextItemComponent = React.forwardRef<HTMLLIElement, IItemProps>(
  ({ children, disabled, ...props }, ref) => {
    const { isCompact } = useMenuContext();

    return (
      <StyledNextItem ref={ref} disabled={disabled} {...props}>
        <StyledItemIcon isCompact={isCompact} isDisabled={disabled} isVisible={true}>
          <StyledNextIcon isDisabled={disabled} />
        </StyledItemIcon>
        {children}
      </StyledNextItem>
    );
  }
);

NextItemComponent.displayName = 'NextItemComponent';

/**
 * Accepts all `<li>` props
 */
export const NextItem = React.forwardRef<HTMLLIElement, Omit<IItemProps, 'component'>>(
  ({ value, disabled, ...props }, ref) => {
    const { nextItemsHashRef } = useDropdownContext();
    const { itemIndexRef } = useMenuContext();

    if (!disabled) {
      // Include current index in global Dropdown context
      (nextItemsHashRef.current as any)[value] = itemIndexRef.current;
    }

    return (
      <Item
        component={NextItemComponent}
        aria-expanded={true}
        disabled={disabled}
        value={value}
        ref={ref}
        {...props}
      />
    );
  }
);

NextItem.displayName = 'NextItem';

NextItem.propTypes = {
  value: PropTypes.any,
  disabled: PropTypes.bool
};
