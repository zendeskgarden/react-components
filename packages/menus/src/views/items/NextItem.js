/**
 * Copyright Zendesk, Inc.
 *
 * Use of this source code is governed under the Apache License, Version 2.0
 * found at http://www.apache.org/licenses/LICENSE-2.0.
 */

import PropTypes from 'prop-types';
import classNames from 'classnames';
import styled from 'styled-components';
import { retrieveComponentStyles } from '@zendeskgarden/react-theming';
import MenuStyles from '@zendeskgarden/css-menus';

import Item from './Item';

const COMPONENT_ID = 'menus.next_item';

/**
 * Accepts all `<li>` props
 */
const NextItem = styled(Item).attrs(props => ({
  'data-garden-id': COMPONENT_ID,
  'data-garden-version': PACKAGE_VERSION,
  className: classNames(props.className, MenuStyles['c-menu__item--next'])
}))`
  ${props => retrieveComponentStyles(COMPONENT_ID, props)};
`;

NextItem.propTypes = {
  active: PropTypes.bool,
  focused: PropTypes.bool,
  hovered: PropTypes.bool,
  disabled: PropTypes.bool
};

NextItem.hasType = () => NextItem;

/** @component */
export default NextItem;
