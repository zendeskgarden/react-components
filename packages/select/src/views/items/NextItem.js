import PropTypes from 'prop-types';
import { retrieveTheme } from '@zendesk/garden-react-theming';
import { NextItem as MenuNextItem } from '@zendesk/garden-react-menus';

import { version } from '../../../package.json';
const COMPONENT_ID = 'select.next_item';

/**
 * Accepts all `<li>` props
 */
const NextItem = MenuNextItem.extend.attrs({
  'data-garden-id': COMPONENT_ID,
  'data-garden-version': version
})`
  ${props => retrieveTheme(COMPONENT_ID, props)};
`;

NextItem.propTypes = {
  active: PropTypes.bool,
  focused: PropTypes.bool,
  hovered: PropTypes.bool,
  disabled: PropTypes.bool
};

/** @component */
export default NextItem;
