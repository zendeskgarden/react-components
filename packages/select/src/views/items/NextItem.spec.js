import React from 'react';
import { shallow } from 'enzyme';
import NextItem from './NextItem';

describe('NextItem', () => {
  it('renders default styling', () => {
    const wrapper = shallow(<NextItem />);

    expect(wrapper).toMatchSnapshot();
  });
});
