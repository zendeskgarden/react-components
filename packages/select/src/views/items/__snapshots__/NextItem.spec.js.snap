// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NextItem renders default styling 1`] = `
<Item
  className="c-menu__item--next "
  data-garden-id="select.next_item"
  data-garden-version="version"
/>
`;
